File: 1.array/121.BestTimeToBuyAndSellStock/main.go
package main

import "fmt"

// https://leetcode.com/problems/remove-element/description/

// input: prices, prices[i]: stock on i day
// maxProfit = prices[j] - prices[i], j: sell day, i: buy day, i < j
// or maxProfit = max(prices) - min(prices)
// At each time (prices[i]), you will sell with the min price in the past to find the maximum of profit
// ex: [7,1,5,3,6,4]
// TC: O(n)
// SC: O(1)

func maxProfit(prices []int) int {
	minPriceSofar := prices[0]
	maxProfit := 0

	for i := 1; i < len(prices); i++ {
		if prices[i] > minPriceSofar {
			maxProfit = max(maxProfit, prices[i]-minPriceSofar)
		}
		minPriceSofar = min(minPriceSofar, prices[i])
	}

	return maxProfit
}

func main() {
	fmt.Println(maxProfit([]int{7, 1, 5, 3, 6, 4}))
}

File: 1.array/121.BestTimeToBuyAndSellStock/main.py
from typing import List

# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

# on each day:
# - tracking the min price in the past
# - calculate profit = current price - min price (in the past)

# TC: O(n), SC: O(1)
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        minPrice = prices[0]
        profit = 0
        for i in range(1, len(prices)):
            profit = max(profit, prices[i] - minPrice)
            minPrice = min(minPrice, prices[i])
        return profit
